java/tools/src/main/java/org/apache/arrow/tools/EchoServer.java

```diff
@@ -66,16 +69,28 @@ public void run() throws IOException {
         // Read the entire input stream.
         reader.init();
         while (true) {
-          ArrowRecordBatch batch = reader.nextRecordBatch();
-          if (batch == null) break;
-          batches.add(batch);
+          Byte type = reader.nextBatchType();
+          if (type == null) {
+            break;
+          } else if (type == MessageHeader.RecordBatch) {
+            batches.add(reader.nextRecordBatch());
+          } else if (type == MessageHeader.DictionaryBatch) {
+            dictionaries.add(reader.nextDictionaryBatch());
+          } else {
+            throw new IOException("Unexpected message header type " + type);
+          }
         }
-        LOGGER.info(String.format("Received %d batches", batches.size()));
+        LOGGER.info(String.format("Received %d batches and %d dictionaries", batches.size(), dictionaries.size()));
 
         // Write it back
         try (ArrowStreamWriter writer = new ArrowStreamWriter(out, reader.getSchema())) {
+          for (ArrowDictionaryBatch batch: dictionaries) {
+            writer.writeDictionaryBatch(batch);
+            batch.close();
+          }
           for (ArrowRecordBatch batch: batches) {
             writer.writeRecordBatch(batch);
+            batch.close();
```

Comment:
I think it was just an oversight - ArrowRecordBatch is Closeable, but they weren't ever being closed here.